### PR TITLE
Make rattr configurable in pyproject toml

### PR DESCRIPTION
**Pull request by habospace**

### CODE REVIEW

**Comment by olliemath** [comment]
Probably a good one for @BPHarris to review :)

**Comment by habospace** [comment]
> Probably a good one for @BPHarris to review :)

Sure, we were discussing it already.

**Comment by habospace** [comment]
Okay, so I feel like before I start to make any tests (and waste a lot of time on writing tests for something that won't even be approved potentially), it would be useful if @BPHarris or somebody could have a look at this and check if this is a good general direction.

**A few notes on the solution:**

Since we're already using the 'argparse' library for cli args, I thought it would be useful if somehow we could reuse our ArgumentParser registries for parsing & validating our .toml cfg args. 
Otherwise, we'd need to create some custom toml arg validation module or find some other kind of toml validator library. I thought those 2 options were suboptimal since making our own toml validator would need a lot of code and testing while some toml validator library would imply that we then have to maintain 2 separate ways of arg validation (for cli and .toml args) at the same time.

This is why I am kind of translating the .toml file into 'sys_argv' style list which can then be parsed and validated using our already defined and tested ArgumentParser (that is set up with the registries).

There was minor issue with the argparse library which is that the 'exit_on_error' param of ArgumentParser doesn't exist for Python versions 3.7 and 3.8. I created a custom/overwritten version of the class which takes this argument and throws an error instead of exiting when 'exit_on_error' is False. It is really important that 'ArgumentParser' is throwing an error, since we want to construct a different error message when error is coming .toml config parsing (otherwise we exit and ArgumentParser has full control over the error message).

**.toml and cli arg merging rule:** 

So the way we merge toml and cli args is we take toml args first, we set every arg that has been set explicitly, everything else stays at default, then we overwrite everything with cli args except in case of lists (args with 'append' action) when we extend .toml args with cli args (so in case of `--exclude-import` if .toml config has  `[a, b, c]` and cli args are `-F d -F e` then the merged `--exlude-imports` are `[a, b, c] + [d, e] = [a, b, c, d, e]`. Let me know if you want a full overwrite here too.

**Comment by habospace** [comment]
@BPHarris  btw, how are you adding "filter_string" in terminal/cli call?

**Comment by habospace** [comment]
actually found a minor bug, let me look into it tomorrow or over the weekend.

**Comment by BPHarris** [comment]
> @BPHarris btw, how are you adding "filter_string" in terminal/cli call?

We don't use it locally in the pipeline, but I did us it in testing like so: `rattr ... my_target_file.py` becomes `rattr ... 'get_*' my_target_file.py`. Is this what you mean?

**Comment by habospace** [comment]
> > @BPHarris btw, how are you adding "filter_string" in terminal/cli call?
> 
> We don't use it locally in the pipeline, but I did us it in testing like so: `rattr ... my_target_file.py` becomes `rattr ... 'get_*' my_target_file.py`. Is this what you mean?

thanks. I think we can go ahead then. ready for review.

File: rattr/cli/parser.py [hunk 1]
```diff
@@ -58,39 +118,123 @@ def parse_arguments() -> Namespace:
...
+        )
+
+    if not project_toml_cfg:
         try:
-            arguments = argument_group_parser.validate(parser, arguments)
+            project_toml_cfg = load_cfg_from_project_toml()
+        except TOMLDecodeError as e:
+            # TODO: maybe construct a more informative message.
+            if exit_on_error:
+                error.fatal("Error parsing pyproject.toml file.")
```
Comment (habospace) [inline rattr/cli/parser.py:160]: @BPHarris  maybe here we should just throw the original error with the original trace (+ maybe a pointer/path to the  .toml file)?

**Review by BPHarris — APPROVED** [review]
Looks good to me :)

**Pull request merged**